src/python/deepgnn/graph_engine/snark/converter/writers.py

```diff
@@ -70,17 +70,33 @@ def __init__(
         self.prev_node_id: typing.Optional[int] = None
         self.prev_edge_type: typing.Optional[int] = None
 
-    def add(self, data: typing.Iterator[typing.Tuple[int, int, int, float, list]]):
+    def add(
+        self,
+        data: typing.Iterator[
+            typing.Tuple[
+                int,
+                int,
+                int,
+                float,
+                list,
+                typing.Optional[int],
+                typing.Optional[int],
+            ]
+        ],
+    ):
         """
         Write binary for a node and all of its edges.
 
         args:
-            data: Iterable[(int, int, int, float, list)] Data for a node first, then all of
+            data: Iterable[(int, int, int, float, list, typing.Optional[typing.Tuple[int, int]])] Data for a node first, then all of
                 its edges in order of type then dst. All of the node's outgoing edges do not need to go to the same writer.
                 Each entry for a node/edge is,
-                (node_id/src, -1/dst, type, weight, [ndarray for each feature vector or None in order of feature index]).
+                (node_id/src, -1/dst, type, weight, [ndarray for each feature vector or None in order of feature index], timestamps of creation and deletion of data).
         """
-        for src, dst, typ, weight, features in data:
+        # Default values for timestamps: node existed since the very beginning and towards the very end.
+        node_created_at = 0
+        node_removed_at = -1
+        for src, dst, typ, weight, features, created_at, removed_at in data:
```

Comment:
Won't this error if created at and removed_at are not specified? Either they should not be marked optional or this is a bug, I think this is a bug.

Reply:
no, tuple has to have 7 values. Optional doesn't affect number of arguments, an individual argument can be None.

Reply:
You can make it so it can have 5 or 7 arguments by unrolling the tuple inside of the for loop instead, this will make it nicer and less confusing for non temporal users.

Reply:
It is already supported for edgelist users through default arguments:
![image](https://user-images.githubusercontent.com/134559/230671994-6597afd5-3b4c-4a91-a14e-8b4c652d2a42.png)

Adding support for variable tuples is not part of this PR, otherwise we should've done this for graphs with same weights/types or without any features.

Reply:
Put it in the changelog then